File: sdk/python/atb/oversight.py
"""Emitters for canonical atb.* oversight event types."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from atb import event_types

__all__ = [
    "ToolCallEmitter",
    "DataExportEmitter",
    "HumanOverrideEmitter",
    "HumanApprovalEmitter",
    "ActionErrorEmitter",
]


def _require_non_empty(value: str, field: str) -> str:
    trimmed = (value or "").strip()
    if not trimmed:
        raise ValueError(f"atb: {field} is required")
    return trimmed


def _dispatch(ctx: Any, event_type: str, payload: Mapping[str, Any]) -> None:
    emit = getattr(ctx, "emit", None)
    if callable(emit):
        emit(event_type, dict(payload))
        return
    append = getattr(ctx, "append", None)
    if callable(append):
        append(event_type, dict(payload))
        return
    raise ValueError("atb: context must provide emit or append")


def _session_id(ctx: Any, kwargs: Mapping[str, Any]) -> str | None:
    override = kwargs.get("session_id")
    if override is not None and str(override).strip():
        return str(override).strip()
    sid = getattr(ctx, "session_id", None)
    if sid is not None and str(sid).strip():
        return str(sid).strip()
    return None


def _base_payload(ctx: Any, kwargs: Mapping[str, Any]) -> dict[str, Any]:
    payload: dict[str, Any] = {}
    sid = _session_id(ctx, kwargs)
    if sid:
        payload["session_id"] = sid
    return payload


def _optional_str(payload: dict[str, Any], kwargs: Mapping[str, Any], key: str) -> None:
    value = kwargs.get(key)
    if value is not None and str(value).strip():
        payload[key] = str(value).strip()


class ToolCallEmitter:
    """Emits atb.tool.call events for direct tool invocations.

    Attach to a WorkflowContext (or compatible event sink) and call emit() once
    per tool invocation that should be recorded in the audit trail.
    """

    def __init__(self, ctx: Any) -> None:
        if not (callable(getattr(ctx, "emit", None)) or callable(getattr(ctx, "append", None))):
            raise ValueError("atb: context must provide emit or append")
        self._ctx = ctx

    def emit(self, tool_name: str, **kwargs: Any) -> None:
        """Emit an atb.tool.call event.

        Args:
            tool_name: The name of the tool invoked. Required.
            **kwargs: Optional fields — actor_id, tool_input_digest,
                tool_output_digest, session_id (overrides context session_id
                if provided).

        Raises:
            ValueError: If tool_name is empty or not provided.
        """
        payload = _base_payload(self._ctx, kwargs)
        payload["tool_name"] = _require_non_empty(tool_name, "tool_name")
        _optional_str(payload, kwargs, "actor_id")
        _optional_str(payload, kwargs, "tool_input_digest")
        _optional_str(payload, kwargs, "tool_output_digest")
        _dispatch(self._ctx, event_types.TOOL_CALL, payload)


class DataExportEmitter:
    """Emits atb.data.export events for outbound data movements."""

    def __init__(self, ctx: Any) -> None:
        if not (callable(getattr(ctx, "emit", None)) or callable(getattr(ctx, "append", None))):
            raise ValueError("atb: context must provide emit or append")
        self._ctx = ctx

    def emit(self, export_target: str, **kwargs: Any) -> None:
        """Emit an atb.data.export event.

        Args:
            export_target: Identifier of the export destination. Required.
            **kwargs: Optional actor_id, record_count (int), classification.

        Raises:
            ValueError: If export_target is empty or not provided.
        """
        payload = _base_payload(self._ctx, kwargs)
        payload["export_target"] = _require_non_empty(export_target, "export_target")
        _optional_str(payload, kwargs, "actor_id")
        record_count = kwargs.get("record_count")
        if isinstance(record_count, int) and record_count > 0:
            payload["record_count"] = record_count
        _optional_str(payload, kwargs, "classification")
        _dispatch(self._ctx, event_types.DATA_EXPORT, payload)


class HumanOverrideEmitter:
    """Emits atb.human.override events when a human overrides an AI recommendation."""

    def __init__(self, ctx: Any) -> None:
        if not (callable(getattr(ctx, "emit", None)) or callable(getattr(ctx, "append", None))):
            raise ValueError("atb: context must provide emit or append")
        self._ctx = ctx

    def emit(self, override_reason: str, **kwargs: Any) -> None:
        """Emit an atb.human.override event.

        Args:
            override_reason: Short rationale for the override. Required.
            **kwargs: Optional actor_id, overridden_action_id.

        Raises:
            ValueError: If override_reason is empty or not provided.
        """
        payload = _base_payload(self._ctx, kwargs)
        payload["override_reason"] = _require_non_empty(override_reason, "override_reason")
        _optional_str(payload, kwargs, "actor_id")
        _optional_str(payload, kwargs, "overridden_action_id")
        _dispatch(self._ctx, event_types.HUMAN_OVERRIDE, payload)


class HumanApprovalEmitter:
    """Emits atb.human.approval events when a human explicitly approves a pending action."""

    def __init__(self, ctx: Any) -> None:
        if not (callable(getattr(ctx, "emit", None)) or callable(getattr(ctx, "append", None))):
            raise ValueError("atb: context must provide emit or append")
        self._ctx = ctx

    def emit(self, approved_action_id: str, **kwargs: Any) -> None:
        """Emit an atb.human.approval event.

        Args:
            approved_action_id: Identifier of the action being approved. Required.
            **kwargs: Optional actor_id, approver_id, note.

        Raises:
            ValueError: If approved_action_id is empty or not provided.
        """
        payload = _base_payload(self._ctx, kwargs)
        payload["approved_action_id"] = _require_non_empty(
            approved_action_id,
            "approved_action_id",
        )
        _optional_str(payload, kwargs, "actor_id")
        _optional_str(payload, kwargs, "approver_id")
        _optional_str(payload, kwargs, "note")
        _dispatch(self._ctx, event_types.HUMAN_APPROVAL, payload)


class ActionErrorEmitter:
    """Emits ai.action.error events for privileged actions that did not succeed."""

    def __init__(self, ctx: Any) -> None:
        if not (callable(getattr(ctx, "emit", None)) or callable(getattr(ctx, "append", None))):
            raise ValueError("atb: context must provide emit or append")
        self._ctx = ctx

    def emit(self, action_id: str, error_class: str, **kwargs: Any) -> None:
        """Emit an ai.action.error event.

        Args:
            action_id: Identifier of the action that failed. Required.
            error_class: One of failed, blocked, timeout, exception,
                denied_at_sink. Required.
            **kwargs: Optional actor_id, error_detail_digest, action_type,
                session_id (overrides context session_id if provided).

        Raises:
            ValueError: If action_id or error_class is empty or not provided.
        """
        payload = _base_payload(self._ctx, kwargs)
        payload["action_id"] = _require_non_empty(action_id, "action_id")
        payload["error_class"] = _require_non_empty(error_class, "error_class")
        _optional_str(payload, kwargs, "actor_id")
        _optional_str(payload, kwargs, "error_detail_digest")
        _optional_str(payload, kwargs, "action_type")
        _dispatch(self._ctx, event_types.AI_ACTION_ERROR_EVENT_TYPE, payload)

File: sdk/typescript/src/oversight.ts
import {
  DATA_EXPORT,
  HUMAN_APPROVAL,
  HUMAN_OVERRIDE,
  TOOL_CALL,
} from "./eventTypes.js";
import { identityEvidencePayload, type IdentityEvidence } from "./identity-evidence.js";

/** Compatible event sink for oversight emitters. */
export interface OversightEventSink {
  emit(type: string, payload: Record<string, unknown>): void;
  sessionId?: string;
  session_id?: string;
}

/** Options for {@link ToolCallEmitter.emit}. */
export interface ToolCallOptions {
  /** Name of the tool invoked. Required. */
  toolName: string;
  actorId?: string;
  /** SHA-256 hex of the canonicalised tool input. */
  inputDigest?: string;
  /** SHA-256 hex of the canonicalised tool output. */
  outputDigest?: string;
}

/** Options for {@link DataExportEmitter.emit}. */
export interface DataExportOptions {
  /** Identifier of the export destination. Required. */
  exportTarget: string;
  actorId?: string;
  recordCount?: number;
  classification?: string;
}

/** Options for {@link HumanOverrideEmitter.emit}. */
export interface HumanOverrideOptions {
  /** Short rationale for the override. Required. */
  overrideReason: string;
  actorId?: string;
  overriddenActionId?: string;
  identityEvidence?: IdentityEvidence;
}

/** Options for {@link HumanApprovalEmitter.emit}. */
export interface HumanApprovalOptions {
  /** Identifier of the action being approved. Required. */
  approvedActionId: string;
  actorId?: string;
  approverId?: string;
  note?: string;
  identityEvidence?: IdentityEvidence;
}

function requireNonEmpty(value: string, field: string): string {
  const trimmed = value?.trim() ?? "";
  if (!trimmed) {
    throw new Error(`atb: ${field} is required`);
  }
  return trimmed;
}

function sessionIdFromCtx(ctx: OversightEventSink): string | undefined {
  const raw = ctx.sessionId ?? ctx.session_id;
  if (raw === undefined || raw === null) {
    return undefined;
  }
  const trimmed = String(raw).trim();
  return trimmed !== "" ? trimmed : undefined;
}

function basePayload(ctx: OversightEventSink): Record<string, unknown> {
  const payload: Record<string, unknown> = {};
  const sid = sessionIdFromCtx(ctx);
  if (sid) {
    payload.session_id = sid;
  }
  return payload;
}

function optionalStr(
  payload: Record<string, unknown>,
  key: string,
  value?: string
): void {
  if (value !== undefined && value.trim() !== "") {
    payload[key] = value.trim();
  }
}

/**
 * Emits atb.tool.call events for direct tool invocations.
 *
 * Attach to a WorkflowContext (or compatible event sink) and call emit() once
 * per tool invocation to be recorded in the audit trail.
 */
export class ToolCallEmitter {
  constructor(private readonly ctx: OversightEventSink) {}

  /**
   * Emit an atb.tool.call event.
   *
   * @param opts Tool call metadata.
   * @throws {Error} If opts.toolName is empty.
   */
  emit(opts: ToolCallOptions): void {
    const payload = basePayload(this.ctx);
    payload.tool_name = requireNonEmpty(opts.toolName, "toolName");
    optionalStr(payload, "actor_id", opts.actorId);
    optionalStr(payload, "tool_input_digest", opts.inputDigest);
    optionalStr(payload, "tool_output_digest", opts.outputDigest);
    this.ctx.emit(TOOL_CALL, payload);
  }
}

/** Emits atb.data.export events for outbound data movements. */
export class DataExportEmitter {
  constructor(private readonly ctx: OversightEventSink) {}

  /**
   * Emit an atb.data.export event.
   *
   * @param opts Export metadata.
   * @throws {Error} If opts.exportTarget is empty.
   */
  emit(opts: DataExportOptions): void {
    const payload = basePayload(this.ctx);
    payload.export_target = requireNonEmpty(opts.exportTarget, "exportTarget");
    optionalStr(payload, "actor_id", opts.actorId);
    if (opts.recordCount !== undefined && opts.recordCount > 0) {
      payload.record_count = opts.recordCount;
    }
    optionalStr(payload, "classification", opts.classification);
    this.ctx.emit(DATA_EXPORT, payload);
  }
}

/**
 * Emits atb.human.override events when a human overrides an AI recommendation
 * or system guardrail.
 */
export class HumanOverrideEmitter {
  constructor(private readonly ctx: OversightEventSink) {}

  /**
   * Emit an atb.human.override event.
   *
   * @param opts Override metadata.
   * @throws {Error} If opts.overrideReason is empty.
   */
  emit(opts: HumanOverrideOptions): void {
    const payload = basePayload(this.ctx);
    payload.override_reason = requireNonEmpty(opts.overrideReason, "overrideReason");
    optionalStr(payload, "actor_id", opts.actorId);
    optionalStr(payload, "overridden_action_id", opts.overriddenActionId);
    const identityEvidence = identityEvidencePayload(opts.identityEvidence);
    if (identityEvidence) {
      payload.identity_evidence = identityEvidence;
    }
    this.ctx.emit(HUMAN_OVERRIDE, payload);
  }
}

/**
 * Emits atb.human.approval events when a human explicitly approves a pending
 * AI action.
 */
export class HumanApprovalEmitter {
  constructor(private readonly ctx: OversightEventSink) {}

  /**
   * Emit an atb.human.approval event.
   *
   * @param opts Approval metadata.
   * @throws {Error} If opts.approvedActionId is empty.
   */
  emit(opts: HumanApprovalOptions): void {
    const payload = basePayload(this.ctx);
    payload.approved_action_id = requireNonEmpty(
      opts.approvedActionId,
      "approvedActionId"
    );
    optionalStr(payload, "actor_id", opts.actorId);
    optionalStr(payload, "approver_id", opts.approverId);
    optionalStr(payload, "note", opts.note);
    const identityEvidence = identityEvidencePayload(opts.identityEvidence);
    if (identityEvidence) {
      payload.identity_evidence = identityEvidence;
    }
    this.ctx.emit(HUMAN_APPROVAL, payload);
  }
}
